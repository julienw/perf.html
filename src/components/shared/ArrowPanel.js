/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import * as React from 'react';
import classNames from 'classnames';

require('./ArrowPanel.css');

type Props = {
  onOpen?: () => mixed,
  onClose?: () => mixed,
  onOkButtonClick?: () => mixed,
  onCancelButtonClick?: () => mixed,
  className: string,
  children: React.Node,
  title?: string,
  okButtonText?: string,
  cancelButtonText?: string,
};

type State = {
  open: boolean,
};

class ArrowPanel extends React.PureComponent<Props, State> {
  _panelElementCreated: (HTMLElement | null) => void;
  _panelElement: HTMLElement | null;

  constructor(props: Props) {
    super(props);
    this.state = { open: false };
    (this: any)._windowMouseDownListener = this._windowMouseDownListener.bind(
      this
    );
    this._panelElementCreated = (elem: HTMLElement | null) => {
      this._panelElement = elem;
    };
    (this: any)._onOkButtonClick = this._onOkButtonClick.bind(this);
    (this: any)._onCancelButtonClick = this._onCancelButtonClick.bind(this);
  }

  open() {
    if (this.state.open) {
      return;
    }

    this.setState({ open: true });
    if (this.props.onOpen) {
      this.props.onOpen();
    }
    window.addEventListener('mousedown', this._windowMouseDownListener, true);
  }

  close() {
    if (!this.state.open) {
      return;
    }

    this.setState({ open: false });
    if (this.props.onClose) {
      this.props.onClose();
    }
    window.removeEventListener(
      'mousedown',
      this._windowMouseDownListener,
      true
    );
  }

  componentWillUnmount() {
    window.removeEventListener(
      'mousedown',
      this._windowMouseDownListener,
      true
    );
  }

  _windowMouseDownListener(e: MouseEvent) {
    const target: Node = (e.target: any); // make flow happy
    if (
      this.state.open &&
      this._panelElement &&
      !this._panelElement.contains(target)
    ) {
      this.close();
    }
  }

  _onOkButtonClick() {
    this.close();
    if (this.props.onOkButtonClick) {
      this.props.onOkButtonClick();
    }
  }

  _onCancelButtonClick() {
    this.close();
    if (this.props.onCancelButtonClick) {
      this.props.onCancelButtonClick();
    }
  }

  render() {
    const {
      className,
      children,
      title,
      okButtonText,
      cancelButtonText,
    } = this.props;
    const hasTitle = title !== undefined;
    const hasButtons = okButtonText || cancelButtonText;
    const { open } = this.state;
    return (
      <div className="arrowPanelAnchor">
        <div
          className={classNames(
            'arrowPanel',
            { open, hasTitle, hasButtons },
            className
          )}
          ref={this._panelElementCreated}
        >
          <div className="arrowPanelArrow" />
          {hasTitle
            ? <h1 className="arrowPanelTitle">
                {title}
              </h1>
            : null}
          <div className="arrowPanelContent">
            {children}
          </div>
          {hasButtons
            ? <div className="arrowPanelButtons">
                <input
                  type="button"
                  className="arrowPanelCancelButton"
                  value={cancelButtonText}
                  onClick={this._onCancelButtonClick}
                />
                <input
                  type="button"
                  className="arrowPanelOkButton"
                  value={okButtonText}
                  onClick={this._onOkButtonClick}
                />
              </div>
            : null}
        </div>
      </div>
    );
  }
}

export default ArrowPanel;
