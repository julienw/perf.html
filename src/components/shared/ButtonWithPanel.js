/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import * as React from 'react';
import classNames from 'classnames';

require('./ButtonWithPanel.css');

type PanelProps = {
  onOpen?: () => mixed,
  onClose?: () => mixed,
};

interface Panel {
  open(): mixed,
}

/**
 * Note about the `panel` prop: we accept any React element whose Component
 * class implements the `Panel` interface above, and has at least the props from
 * `PanelProps` above, and any State type. */
type Props = {
  className: string,
  label: string,
  panel: React.Element<
    Class<Panel & React.Component<$Subtype<PanelProps>, any>>
  >,
  open?: boolean,
  disabled?: boolean,
};

type State = {|
  open: boolean,
|};

class ButtonWithPanel extends React.PureComponent<Props, State> {
  _panel: Panel | null;

  _onPanelOpen: () => void;
  _onPanelClose: () => void;
  _panelCreated: (Panel | null) => void;

  constructor(props: Props) {
    super(props);
    this.state = { open: !!props.open };
    this._onPanelOpen = () => {
      this.setState({ open: true });
      if (this.props.panel.props.onOpen) {
        this.props.panel.props.onOpen();
      }
    };
    this._onPanelClose = () => {
      this.setState({ open: false });
      if (this.props.panel.props.onClose) {
        this.props.panel.props.onClose();
      }
    };
    (this: any)._onButtonClick = this._onButtonClick.bind(this);
    this._panelCreated = (panel: Panel | null) => {
      this._panel = panel;
    };
  }

  componentWillReceiveProps(props: Props) {
    if (props.open !== this.props.open) {
      this.setState({ open: !!props.open });
    }
  }

  openPanel() {
    if (this._panel) {
      this._panel.open();
    }
  }

  _onButtonClick() {
    this.openPanel();
  }

  render() {
    const { className, label, panel, disabled } = this.props;
    const { open } = this.state;
    return (
      <div className={classNames('buttonWithPanel', className, { open })}>
        <div className="buttonWithPanelButtonWrapper">
          <input
            type="button"
            className={classNames(
              'buttonWithPanelButton',
              `${className}Button`
            )}
            disabled={!!disabled}
            value={label}
            onClick={this._onButtonClick}
          />
        </div>
        {React.cloneElement(panel, {
          ref: this._panelCreated,
          onOpen: this._onPanelOpen,
          onClose: this._onPanelClose,
        })}
      </div>
    );
  }
}

export default ButtonWithPanel;
