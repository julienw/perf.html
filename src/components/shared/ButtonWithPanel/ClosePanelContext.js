/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

// This is a very simple context that's used for communication between the panel
// and its content, so that it can be closed from actions in inner content.

import * as React from 'react';

export const ClosePanelContext: React.Context<
  (void) => mixed
> = React.createContext(() => {});
