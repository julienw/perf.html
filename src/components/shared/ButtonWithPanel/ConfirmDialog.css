/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

.confirmDialogTitle {
  margin: 0;
  font-size: 1.2em;
}

.confirmDialogContent {
  margin: 16px 0;
}

.confirmDialogButtons {
  display: flex;
  gap: 8px;
}
