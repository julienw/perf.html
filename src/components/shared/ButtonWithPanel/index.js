/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

// This implements a Button that triggers a panel. The panel will have a small
// arrow pointing towards the button, which is implemented in ArrowPanel.
//
// All public API is exposed from this file. Other components should never be
// used directly.
//
// Here is a simple example:
//
// import { ButtonWithPanel } from // 'firefox-profiler/components/ButtonWithPanel';
// ...
// <ButtonWithPanel
//   className="MyButtonWithPanel"
//   buttonClassName="MyPanelButton"
//   panelClassName="MyPanel"
//   label="Click me!"
//   panelContent={<>
//     <p>We explain lots of useful things here.</p>
//     <p>If you want to know more <a href='/'>click here</a>.</p>
//   </>}
// />

export { ButtonWithPanel } from './ButtonWithPanel';
