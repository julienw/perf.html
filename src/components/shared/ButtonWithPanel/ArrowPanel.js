/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

// This implements a panel with a "arrow" triangle graphic that points to the
// button that triggers it.
// Please do not use this component directly. This is used by ButtonWithPanel
// to wrap the panel content.

import * as React from 'react';
import classNames from 'classnames';

import './ArrowPanel.css';

type Props = {|
  +onOpen: () => mixed,
  +onClose: () => mixed,
  +className?: string,
  +children: React.Node,
|};

type State = {|
  +open: boolean,
  +isClosing: boolean,
  +openGeneration: number,
|};

export class ArrowPanel extends React.PureComponent<Props, State> {
  state = {
    open: false,
    isClosing: false,
    openGeneration: 0,
  };

  open() {
    if (this.state.open) {
      return;
    }

    this.setState({ open: true });
    this.props.onOpen();
  }

  close() {
    this.setState(state => {
      if (!state.open) {
        return null;
      }
      const openGeneration = state.openGeneration + 1;

      setTimeout(this._onCloseAnimationFinish(openGeneration), 400);

      this.props.onClose();

      return { open: false, isClosing: true, openGeneration };
    });
  }

  _onCloseAnimationFinish(openGeneration: number) {
    return () => {
      this.setState(state => {
        if (state.openGeneration === openGeneration) {
          return { isClosing: false };
        }
        return null;
      });
    };
  }

  _onArrowPanelClick = (e: { target: HTMLElement } & SyntheticMouseEvent<>) => {
    // The arrow panel element contains the element that has the top arrow,
    // that is visually outside the panel. We still want to hide the panel
    // when clicking in this area.
    if (e.target.className !== 'arrowPanelArrow') {
      // Stop the click propagation to reach the _onWindowClick event when the
      // click is visually inside the panel.
      e.stopPropagation();
    }
  };

  render() {
    const { className, children } = this.props;
    const { open, isClosing } = this.state;
    if (!open && !isClosing) {
      return null;
    }

    return (
      <div className="arrowPanelAnchor">
        <div
          className={classNames('arrowPanel', { open }, className)}
          onClick={this._onArrowPanelClick}
        >
          <div className="arrowPanelArrow" />
          <div className="arrowPanelContent">{children}</div>
        </div>
      </div>
    );
  }
}
