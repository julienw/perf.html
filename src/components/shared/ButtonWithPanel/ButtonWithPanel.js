/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

// This implements the Button, as well as registering the events on the window
// object (mouse and keyboard events).

import * as React from 'react';
import classNames from 'classnames';

import { ArrowPanel } from './ArrowPanel';

import './ButtonWithPanel.css';

type Props = {|
  +className?: string,
  +label: string,
  +panelContent: React.Node,
  +panelClassName?: string,
  // This prop tells the panel to be open by default, but the open/close state is fully
  // managed by the ButtonWithPanel component.
  +initialOpen?: boolean,
  // The class name of the button input element.
  +buttonClassName?: string,
  +onPanelOpen?: () => mixed,
  +onPanelClose?: () => mixed,
  +buttonRef?: {| -current: null | HTMLInputElement |},
|};

type State = {|
  +open: boolean,
|};

export class ButtonWithPanel extends React.PureComponent<Props, State> {
  _panel: ArrowPanel | null = null;

  constructor(props: Props) {
    super(props);
    this.state = { open: !!props.initialOpen };
  }

  componentDidMount() {
    // the panel can be closed by clicking anywhere on the window
    window.addEventListener('click', this._onWindowClick);
    // the panel can be closed by pressing the Esc key
    window.addEventListener('keydown', this._onKeyDown);
    if (this.state.open) {
      this.openPanel();
    }
  }

  componentWillUnmount() {
    window.removeEventListener('keydown', this._onKeyDown);
    window.removeEventListener('click', this._onWindowClick);
  }

  _onPanelOpen = () => {
    this.setState({ open: true });
    if (this.props.onPanelOpen) {
      this.props.onPanelOpen();
    }
  };

  _onPanelClose = () => {
    this.setState({ open: false });
    if (this.props.onPanelClose) {
      this.props.onPanelClose();
    }
  };

  _takePanelRef = (panel: ArrowPanel | null) => {
    this._panel = panel;
  };

  openPanel() {
    if (this._panel) {
      this._panel.open();
    }
  }

  closePanel() {
    if (this._panel && this.state.open) {
      this._panel.close();
    }
  }

  _onWindowClick = () => {
    this.closePanel();
  };

  _onButtonClick = () => {
    if (!this.state.open) {
      // We use a timeout so that we let the event bubble up to the handlers bound
      // on `window`, closing all other panels, before opening this one.
      setTimeout(() => this.openPanel());
    }
  };

  _onKeyDown = (e: KeyboardEvent) => {
    if (e.key === 'Escape') {
      this.closePanel();
    }
  };

  render() {
    const {
      className,
      label,
      panelContent,
      panelClassName,
      buttonClassName,
      buttonRef,
    } = this.props;
    const { open } = this.state;
    return (
      <div className={classNames('buttonWithPanel', className, { open })}>
        <input
          type="button"
          className={classNames('buttonWithPanelButton', buttonClassName)}
          value={label}
          onClick={this._onButtonClick}
          ref={buttonRef}
        />
        <ArrowPanel
          className={panelClassName}
          onOpen={this._onPanelOpen}
          onClose={this._onPanelClose}
          ref={this._takePanelRef}
        >
          {panelContent}
        </ArrowPanel>
      </div>
    );
  }
}
