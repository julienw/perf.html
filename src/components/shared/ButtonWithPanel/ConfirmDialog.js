/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import * as React from 'react';
import classNames from 'classnames';

import { ClosePanelContext } from './ClosePanelContext';

import './ConfirmDialog.css';

type Props = {|
  // If these callbacks return promises, they will be waited for before the
  // panel is closed. To make this explicit we added Promise<mixed> even if that
  // isn't necessary to pass Flow checks.
  +onCancelButtonClick?: () => mixed | Promise<mixed>,
  +onConfirmButtonClick?: () => mixed | Promise<mixed>,
  +className?: string,
  +title: string,
  +cancelButtonText?: string,
  +confirmButtonText?: string,
  +confirmButtonType?: 'default' | 'primary' | 'destructive',
  +children: React.Node,
|};

type PropsWithClose = {|
  ...Props,
  closePanel: void => mixed,
|};

type State = {|
  status: 'idle' | 'working',
|};

export class ConfirmDialogImpl extends React.PureComponent<
  PropsWithClose,
  State
> {
  state = {
    status: 'idle',
  };

  _onConfirmButtonClick = async () => {
    if (this.state.status === 'working') {
      return;
    }

    this.setState({ status: 'working' });
    if (this.props.onConfirmButtonClick) {
      try {
        await this.props.onConfirmButtonClick();
      } finally {
        this.setState({ status: 'idle' });
      }
    }
    this.props.closePanel();
  };

  _onCancelButtonClick = async () => {
    if (this.state.status === 'working') {
      return;
    }

    this.setState({ status: 'working' });
    if (this.props.onCancelButtonClick) {
      try {
        await this.props.onCancelButtonClick();
      } finally {
        this.setState({ status: 'idle' });
      }
    }
    this.props.closePanel();
  };

  render() {
    const {
      className,
      children,
      title,
      cancelButtonText,
      confirmButtonText,
      confirmButtonType,
    } = this.props;
    const { status } = this.state;

    return (
      <div className={classNames('confirmDialog', className)}>
        <h2 className="confirmDialogTitle">{title}</h2>
        <div className="confirmDialogContent">{children}</div>
        <div className="confirmDialogButtons">
          <input
            type="button"
            className="photon-button photon-button-default"
            value={cancelButtonText || 'Cancel'}
            disabled={status === 'working'}
            onClick={this._onCancelButtonClick}
          />
          <input
            type="button"
            className={`photon-button photon-button-${confirmButtonType ||
              'primary'}`}
            value={confirmButtonText || 'Confirm'}
            disabled={status === 'working'}
            onClick={this._onConfirmButtonClick}
          />
        </div>
      </div>
    );
  }
}

export function ConfirmDialog(props: Props) {
  return (
    <ClosePanelContext.Consumer>
      {panelCloseFunction => (
        <ConfirmDialogImpl {...props} closePanel={panelCloseFunction} />
      )}
    </ClosePanelContext.Consumer>
  );
}
