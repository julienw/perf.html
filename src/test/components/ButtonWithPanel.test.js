/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import * as React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { ButtonWithPanel } from '../../components/shared/ButtonWithPanel';
import { ensureExists } from '../../utils/flow';
import { fireFullClick } from '../fixtures/utils';

beforeEach(() => {
  jest.useFakeTimers();
});

describe('shared/ButtonWithPanel', () => {
  // renders the button in its default state
  function setup() {
    return render(
      <ButtonWithPanel
        className="button"
        buttonClassName="buttonButton"
        label="My Button"
        panelClassName="panel"
        panelContent={<div>Panel content</div>}
      />
    );
  }

  it('renders the ButtonWithPanel with a closed panel', () => {
    const { container } = setup();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders a button with a panel', () => {
    const { container } = render(
      <ButtonWithPanel
        className="button"
        label="My Button"
        initialOpen={true}
        panelClassName="panel"
        panelContent={<div>Panel content</div>}
      />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  /* These tests will be rewritten when ConfirmDialog is introduced.
  describe('various panel contents', () => {
    it('renders panels with default buttons and titles', () => {
      const { container } = render(
        <ButtonWithPanel
          className="button"
          label="My Button"
          initialOpen={true}
          panel={
            <ArrowPanel
              className="panel"
              title="Wonderful content"
              okButtonText="Confirm"
              cancelButtonText="Cancel"
            >
              <div>Panel content</div>
            </ArrowPanel>
          }
        />
      );
      expect(container.firstChild).toMatchSnapshot();
    });

    it('renders panels with specified buttons', () => {
      const { container } = render(
        <ButtonWithPanel
          className="button"
          label="My Button"
          initialOpen={true}
          panel={
            <ArrowPanel
              className="panel"
              okButtonText="Confirm"
              okButtonType="destructive"
              cancelButtonText="Cancel"
            >
              <div>Panel content</div>
            </ArrowPanel>
          }
        />
      );
      expect(container.firstChild).toMatchSnapshot();
    });
  });
  */

  describe('protecting against mounting expensive panels', function() {
    it('does not render the contents when closed', function() {
      const { queryByTestId } = render(
        <ButtonWithPanel
          className="button"
          label="My Button"
          panelContent={<div data-testid="panel-content">Panel content</div>}
        />
      );
      expect(queryByTestId('panel-content')).toBeFalsy();
    });

    /**
     * This test asserts that we don't try and mount the contents of a panel if it's
     * not open.
     */
    it('only renders the contents when open', function() {
      const { queryByTestId } = render(
        <ButtonWithPanel
          className="button"
          label="My Button"
          initialOpen={true}
          panelContent={<div data-testid="panel-content">Panel content</div>}
        />
      );
      expect(queryByTestId('panel-content')).toBeTruthy();
    });
  });

  it('opens the panel when the button is clicked and closes the panel when the escape key is pressed', () => {
    const { getByText, container } = setup();

    fireFullClick(getByText('My Button'));
    jest.runAllTimers();
    expect(container.firstChild).toMatchSnapshot();

    //it closes the panel when Esc key is pressed
    fireEvent.keyDown(container, {
      key: 'Escape',
      keyCode: 27,
      which: 27,
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  it('opens the panel when the button is clicked and closes the panel by clicking outside the panel', () => {
    const { getByText, container } = setup();

    fireFullClick(getByText('My Button'));
    jest.runAllTimers();
    expect(container.firstChild).toMatchSnapshot();

    // it closes the panel when clicking outside the panel
    const newDiv = ensureExists(document.body).appendChild(
      document.createElement('div')
    );
    fireFullClick(newDiv);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('opens the panel when the button is clicked and closes the panel by clicking the button again', () => {
    const { getByText, container } = setup();

    fireFullClick(getByText('My Button'));
    jest.runAllTimers();

    ensureExists(container.querySelector('.arrowPanel.open'));

    fireFullClick(getByText('My Button'));
    jest.runAllTimers();

    expect(container.querySelector('.arrowPanel.open')).toBe(null);
  });

  it('opens the panel when the button is clicked and does not close the panel by clicking inside the panel', () => {
    const { getByText, container } = setup();

    fireFullClick(getByText('My Button'));
    jest.runAllTimers();
    ensureExists(container.querySelector('.arrowPanel.open'));

    // Clicking on the panel doesn't hide the popup.
    fireFullClick(getByText('Panel content'));
    jest.runAllTimers();
    ensureExists(container.querySelector('.arrowPanel.open'));

    // But clicking on the arrow area does.
    fireFullClick(ensureExists(container.querySelector('.arrowPanelArrow')));
    jest.runAllTimers();
    expect(container.querySelector('.arrowPanel.open')).toBe(null);
  });
});
