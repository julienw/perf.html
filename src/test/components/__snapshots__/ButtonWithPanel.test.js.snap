// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shared/ButtonWithPanel opens the panel when the button is clicked and closes the panel by clicking outside the panel 1`] = `
<div
  class="buttonWithPanel button open"
>
  <input
    class="buttonWithPanelButton buttonButton"
    type="button"
    value="My Button"
  />
  <div
    class="arrowPanelAnchor"
  >
    <div
      class="arrowPanel open panel"
    >
      <div
        class="arrowPanelArrow"
      />
      <div
        class="arrowPanelContent"
      >
        <div>
          Panel content
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`shared/ButtonWithPanel opens the panel when the button is clicked and closes the panel by clicking outside the panel 2`] = `
<div
  class="buttonWithPanel button"
>
  <input
    class="buttonWithPanelButton buttonButton"
    type="button"
    value="My Button"
  />
  <div
    class="arrowPanelAnchor"
  >
    <div
      class="arrowPanel panel"
    >
      <div
        class="arrowPanelArrow"
      />
      <div
        class="arrowPanelContent"
      >
        <div>
          Panel content
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`shared/ButtonWithPanel opens the panel when the button is clicked and closes the panel when the escape key is pressed 1`] = `
<div
  class="buttonWithPanel button open"
>
  <input
    class="buttonWithPanelButton buttonButton"
    type="button"
    value="My Button"
  />
  <div
    class="arrowPanelAnchor"
  >
    <div
      class="arrowPanel open panel"
    >
      <div
        class="arrowPanelArrow"
      />
      <div
        class="arrowPanelContent"
      >
        <div>
          Panel content
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`shared/ButtonWithPanel opens the panel when the button is clicked and closes the panel when the escape key is pressed 2`] = `
<div
  class="buttonWithPanel button"
>
  <input
    class="buttonWithPanelButton buttonButton"
    type="button"
    value="My Button"
  />
  <div
    class="arrowPanelAnchor"
  >
    <div
      class="arrowPanel panel"
    >
      <div
        class="arrowPanelArrow"
      />
      <div
        class="arrowPanelContent"
      >
        <div>
          Panel content
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`shared/ButtonWithPanel renders a button with a panel 1`] = `
<div
  class="buttonWithPanel button open"
>
  <input
    class="buttonWithPanelButton"
    type="button"
    value="My Button"
  />
  <div
    class="arrowPanelAnchor"
  >
    <div
      class="arrowPanel open panel"
    >
      <div
        class="arrowPanelArrow"
      />
      <div
        class="arrowPanelContent"
      >
        <div>
          Panel content
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`shared/ButtonWithPanel renders the ButtonWithPanel with a closed panel 1`] = `
<div
  class="buttonWithPanel button"
>
  <input
    class="buttonWithPanelButton buttonButton"
    type="button"
    value="My Button"
  />
</div>
`;
